demo/src/server.rs

```diff
@@ -0,0 +1,277 @@
+    ) -> Result<Self, ServerError<'l>> {
+        let engine = Engine::new(
+            language_set,
+            notation_sets,
+            make_example_pane_notation(),
+            make_keyhint_lang(),
+            make_message_lang(),
+            make_json_lang(),
+        )?;
+        let mut keymap_manager = KeymapManager::new();
+        keymap_manager.register_mode("tree".into(), example_keymaps::make_tree_map());
+        keymap_manager.register_mode("speed_bool".into(), example_keymaps::make_speed_bool_map());
+        keymap_manager.register_menu(
+            "node".into(),
+            example_keymaps::make_node_map(
+                engine.language(engine.lang_name_of(&DocLabel::ActiveDoc)?)?,
+            ),
+        );
+        keymap_manager.replace_text_keymap(example_keymaps::make_text_map());
+
+        let mut ed = Server {
+            engine,
+            frontend: Terminal::new(ColorTheme::default_dark())?,
+            data_stack: DataStack::new(),
+            call_stack: CallStack::new(),
+            keymap_manager,
+        };
+
+        // Set initial keymap
+        ed.call(Word::Literal(Value::ModeName("tree".into())))?;
+        ed.call(Word::PushMode)?;
+
+        // ed.engine.clear_messages()?;
+        Ok(ed)
+    }
+
+    pub fn run(&mut self) -> Result<(), ServerError<'l>> {
+        loop {
+            if self.keymap_manager.has_active_menu() {
+                self.handle_input()?;
+            } else {
+                if let Some(word) = self.call_stack.next() {
+                    if let Err(err) = self.call(word) {
+                        self.engine.show_message(&format!("Error: {}", err))?;
+                    }
+                } else {
+                    self.engine.exec(MetaCommand::EndGroup)?;
+                    self.handle_input()?;
+                }
+            }
+        }
+    }
+
+    fn handle_input(&mut self) -> Result<(), ServerError<'l>> {
+        let doc = self.engine.active_doc()?;
+        let tree_context = if doc.in_tree_mode() {
+            Some(FilterContext {
+                required_sort: doc.self_sort(),
+                self_arity: doc.self_arity_type(),
+                parent_arity: doc.parent_arity_type(),
+            })
+        } else {
+            None
+        };
+        let keymap = self.keymap_manager.get_active_keymap(tree_context)?;
+
+        self.update_key_hints(&keymap)?;
+        self.engine.redisplay(&mut self.frontend)?;
+        match self.next_event(&keymap) {
+            Ok(prog) => {
+                self.call_stack.push(prog);
+                self.keymap_manager.deactivate_menu();
+                Ok(())
+            }
+            Err(ServerError::KeyboardInterrupt) => Err(ServerError::KeyboardInterrupt),
+            Err(err) => Ok(self.engine.show_message(&format!("Error: {}", err))?),
+        }
+    }
+
+    fn update_key_hints(&mut self, keymap: &FilteredKeymap) -> Result<(), ServerError<'l>> {
```

Comment:
For the future: make a "Virtual Doc" that implements PrettyDocument, b.c. ASTs have a terrible API for this use case, as demonstrated in this method's code. (Add a TODO.)